qutip/core/cy/qobjevo.pyx

```diff
@@ -0,0 +1,833 @@
+        binop between QobjEvo with different shift.
+        """
+        cdef QobjEvo out = self.copy()
+        out._shift_dt = dt
+        return out
+
+    def tidyup(self, atol=1e-12):
+        """Removes small elements from quantum object."""
+        for element in self.elements:
+            if type(element) is _ConstantElement:
+                element = _ConstantElement(element.qobj(0).tidyup(atol))
+            elif type(element) is _EvoElement:
+                element = _EvoElement(element.qobj(0).tidyup(atol),
+                                      element._coefficient)
+        return self
+
+    def linear_map(self, op_mapping, *, _skip_check=False):
+        """
+        Apply mapping to each Qobj contribution.
+
+        Example:
+        `QobjEvo([sigmax(), coeff]).linear_map(spre)`
+        gives the same result has
+        `QobjEvo([spre(sigmax()), coeff])`
+
+        Returns
+        -------
+        :class:`.QobjEvo`
+            Modified object
+
+        Notes
+        -----
+        Does not modify the coefficients, thus `linear_map(conj)` would not
+        give the the conjugate of the QobjEvo. It's only valid for linear
+        transformations.
+        """
+        if not _skip_check:
+            out = op_mapping(self(0))
+            if not isinstance(out, Qobj):
+                raise TypeError("The op_mapping function must return a Qobj")
+        cdef QobjEvo res = self.copy()
+        res.elements = [element.linear_map(op_mapping) for element in res.elements]
+        res.dims = res.elements[0].qobj(0).dims
+        res.shape = res.elements[0].qobj(0).shape
+        res._issuper = res.elements[0].qobj(0).issuper
+        res._isoper = res.elements[0].qobj(0).isoper
+        if not _skip_check:
+            if res(0) != out:
+                raise ValueError("The mapping is not linear")
+
+        return res
+
+    ###########################################################################
+    # Cleaning and compress                                                   #
+    ###########################################################################
+    def _compress_merge_qobj(self, coeff_elements):
+        "merge element with matching qobj: [A, f1], [A, f2] -> [A, f1+f2]"
+        cleaned_elements = []
+        # Mimic a dict with Qobj not hashable
+        qobjs = []
+        coeffs = []
+        for element in coeff_elements:
+            for i, qobj in enumerate(qobjs):
+                if element.qobj(0) == qobj:
+                    coeffs[i] = coeffs[i] + element._coefficient
+                    break
+            else:
+                qobjs.append(element.qobj(0))
+                coeffs.append(element._coefficient)
+        for qobj, coeff in zip(qobjs, coeffs):
+            cleaned_elements.append(_EvoElement(qobj, coeff))
+        return cleaned_elements
+
+    def compress(self):
+        """
+        Look for redundance in the QobjEvo components, then merge them.
```

Comment:
Could we extend this document with a list of what optimizations currently occur?